### PR TITLE
Jaime Mitchell- AdagramsJS - D18 - SnowLeopards

### PR DESCRIPTION
**Pull request by JaimeMitchell**
I added a couple of edge case tests.

### CODE REVIEW

File: test/adagrams.test.js [hunk 1]
```diff
@@ -3,7 +3,7 @@ import {
   usesAvailableLetters,
   scoreWord,
   highestScoreFrom,
-} from "adagrams";
+} from 'adagrams';
```
Comment (anselrognlie) [inline test/adagrams.test.js:6]: Argh! prettier!!! 😂

File: test/adagrams.test.js [hunk 2]
```diff
@@ -70,60 +70,60 @@ describe("Adagrams", () => {
...
         a: 1,
         dog: 5,
         wHiMsY: 17,
       });
     });
 
-    it("returns a score of 0 if given an empty input", () => {
-      throw "Complete test";
+    it('returns a score of 0 if given an empty input', () => {
+      expectScores({});
```
Comment (anselrognlie) [inline test/adagrams.test.js]: 👀 `expectScores` (implementation above) works by iterating over the entries in the supplied object, treating the key as the word, and the value as the expected score to check for.  It's not the direct input to our function. So this passes, but that's because we didn't ask `expectScores` to check anything. To have it key that an empty string has a score of 0, we would do something like
```js
      expectScores({
        "": 0
      });
```

It may be surprising that we can use an empty string as an object key, but even an empty string is a string.

File: test/adagrams.test.js [hunk 3]
```diff
@@ -133,22 +133,22 @@ describe("Adagrams", () => {
...
 
-    it("accurately finds best scoring word even if not sorted", () => {
-      const words = ["XXX", "XXXX", "X", "XX"];
-      const correct = { word: "XXXX", score: scoreWord("XXXX") };
+    it('accurately finds best scoring word even if not sorted', () => {
+      const words = ['XXX', 'XXXX', 'X', 'XX'];
+      const correct = { word: 'XXXX', score: scoreWord('XXXX') };
 
-      throw "Complete test by adding an assertion";
+      expect(highestScoreFrom(words)).toEqual(correct);
```
Comment (anselrognlie) [inline test/adagrams.test.js:148]: 👍

File: src/adagrams.js [hunk 4]
```diff
@@ -1,15 +1,173 @@
 export const drawLetters = () => {
   // Implement this method for wave 1
+  let hand = [];
+  let letterBag = [];
+  const letterObj = {
+    A: 9,
```
Comment (anselrognlie) [inline src/adagrams.js:6]: Why are the letters in this ordering? It makes it hard to tell whether we have all the letters represented.

File: src/adagrams.js [hunk 5]
```diff
@@ -1,15 +1,173 @@
...
+  //we need 10 letters
+  let pickedLetter;
+  for (let i = 0; i < 10; i++) {
+    //logic of pop, since i can't pass an argument to it:
+    //1. random pick from handDict   // minus one from letterDict
+    let index = Math.floor(Math.random() * letterBag.length);
+    //2. use the random index in letterbag to grab it.
+    pickedLetter = letterBag[index];
+    //3. remove it from the bag
+    letterBag.splice(index, 1);
```
Comment (anselrognlie) [inline src/adagrams.js:48]: Yes, this is how we would "spell" python's version of `pop` in JS. Keep in mind that this will need to shift the contents of the array, meaning this is a O(n) call that we're making within a loop. Since we're working with fixed lengths (26 letters a certain number of times each, and a 10 letter hand), we don't need to worry too much about the absolute performance, but in general we'd like to find approaches that don't perform linear complexity operations within a loop.

File: test/adagrams.test.js [hunk 6]
```diff
@@ -158,42 +158,68 @@ describe("Adagrams", () => {
...
+      it('selects the word with fewer letters when neither are 10 letters', () => {
+        const words = ['MMMM', 'WWW'];
+        const correct = { word: 'WWW', score: scoreWord('WWW') };
         expectTie(words);
 
         expect(highestScoreFrom(words)).toEqual(correct);
         expect(highestScoreFrom(words.reverse())).toEqual(correct);
       });
+      //I added this because my game logic is set up to return the '10 letters' and the very first one, which might fail under tie words under 10.
+      it('selects the first word when both have same length of 9', () => {
```
Comment (anselrognlie) [inline test/adagrams.test.js:182]: 👍 Never a bad idea to add additional tests when there's a part of your code you're uncertain of.

File: test/adagrams.test.js [hunk 7]
```diff
@@ -158,42 +158,68 @@ describe("Adagrams", () => {
...
         };
         expectTie(words);
 
         expect(highestScoreFrom(words)).toEqual(first);
         expect(highestScoreFrom(words.reverse())).toEqual(second);
       });
+
+      //check multiple tie words (since the others are just 2 words and sorted (ascending and descending) as a result which may let bugs fall through with tie arrays that are greater than 2, unsorted)
+      it('selects the correct word of multiple words and unsorted to check tie logic for edge cases', () => {
+        const words = ['XD', 'Z', 'AAAAFG', 'KDAAA'];
```
Comment (anselrognlie) [inline test/adagrams.test.js:217]: `AAAAFG` sounds like a cry to the heavens 😆

File: src/adagrams.js [hunk 8]
```diff
@@ -1,15 +1,173 @@
...
+
+  return hand;
 };
 
 export const usesAvailableLetters = (input, lettersInHand) => {
-  // Implement this method for wave 2
+  //iterate over each letter, if letter in lettersInHand remove it from lettersInHand to make sure it can't be used again.
+  for (const l of input) {
+    if (lettersInHand.includes(l)) {
+      lettersInHand.splice(l, 1);
```
Comment (anselrognlie) [inline src/adagrams.js]: 👀 `splice` takes a _position_ from which to start removing data, but `l` here is the letter we're iterating over in the input word. The code appears to function, because JS tries to convert the supplied value to a number using `Number()` to do so. `Number("a")` return `NaN`. And `splice` is silently interpretting that as a `0`. 

The tests all pass because the "important" letters are all in order at the beginning of the had, so as the letters are pared off the front of the hand with each iteration, it all hapens to line up in such a way that each test happens to pass.

Comment (anselrognlie) [reply src/adagrams.js]: Observe that we are performing operations (`includes` and `splice`) which require iterating through the array of data, and that we are in a loop. So we're iterating over the data each time of the outer loop.

Consider building frequency maps for the input and the hand which requires iterating over each only once. Then rather than repeatedly removing letters from the hand array, we could compare the counts of each input letter with the available count in the hand.

File: src/adagrams.js [hunk 9]
```diff
@@ -1,15 +1,173 @@
...
+  //returns true because it went through full loop without returning out
+  return true;
 };
 
 export const scoreWord = (word) => {
-  // Implement this method for wave 3
+  //wave3
+  // convert word to uppercase
+  word = word.toUpperCase();
+  const scoreObj = {
```
Comment (anselrognlie) [inline src/adagrams.js:76]: Consider moving this outside the function (or wrap it aggressively) so that the function becomes shorter.

File: src/adagrams.js [hunk 10]
```diff
@@ -1,15 +1,173 @@
...
+    J: 8,
+    X: 8,
+    Q: 10,
+    Z: 10,
+  };
+  let score = 0;
+  //guard condition returns score
+  if (typeof word !== 'string' || word === '') {
+    return score;
+  }
```
Comment (anselrognlie) [inline src/adagrams.js:108]: We didn't have any requirement to validate the type of the input. The earlier call to `toUpperCase` would have crashed if it _weren't_ a string, so we can omit the `type` check here (since we already yolo'd it) 😁 

The empty string check is also unnecesary. If the word is empty, `score` will be initialized to 0, the loop will not be entered (leaving it 0), and no bonus will be added, still leaving it 0.

File: src/adagrams.js [hunk 11]
```diff
@@ -1,15 +1,173 @@
...
+    // }
+  }
+
+  //use highscore and low in a compound conditional that checks for high score
+  //lowest letter num && any that equal it equality && obj.word.length === 10, then append them
+  for (const obj of wordArr) {
+    if (obj.score === highScore.score && obj !== highScore) {
+      tieArr.push(obj);
+    }
+    tieArr.push(highScore);
```
Comment (anselrognlie) [inline src/adagrams.js]: This will end up pushing `wordArr.length` copies of `highScore` into `tieArr`. You'd just want one copy, and it should probably appear first in the list, since it would have been the first item found with the high score. I think you could just get away without the identity check on 144, since we do want all objects having the high score in the eventual tie list. So there's a little extra going on here, but I think the ordering ends up being preserved.

File: src/adagrams.js [hunk 12]
```diff
@@ -1,15 +1,173 @@
...
+  }
+  //TIE LOGIC
+  if (tieArr.length > 1) {
+    for (const obj of tieArr) {
+      if (obj.word.length === 10) {
+        return (highScore = obj);
```
Comment (anselrognlie) [inline src/adagrams.js]: `return obj;` on its own would be fine. We don't need to update `highScore` since we're returning.

File: src/adagrams.js [hunk 13]
```diff
@@ -1,15 +1,173 @@
...
+      }
+
+      if (obj.word.length === low.word.length) {
+        highScore = low;
+      }
```
Comment (anselrognlie) [inline src/adagrams.js]: I don't think this check is necessary. It says if the new length is the same as the shortest length so far, keep the shortest so far, so there's no observable code effect.

**Review by anselrognlie — COMMENTED** [review]
Nice work getting situated in JS. Remember to keep in mind where certain built-in function calls might be hiding a loop, and try to avoid using those within an outer loop (things like `splice` or `includes`). Even though the small size of the data in this means the absolute performance is still fast, we should keep the algorithmic complexity in mind. Imagine what would happen if we relaxed the data restrictions to use larger tile sets or hands.

Take a look at my feedback on `usesAvailableLetters` and the test `returns a score of 0 if given an empty input`, for a few important points, but otherwise nicely done.
